diff --git a/README.md b/README.md
index 1a068fa..2d4cb51 100644
--- a/README.md
+++ b/README.md
@@ -370,6 +370,19 @@ That is, a pattern like `a///b` will match the file path `a/b`.
 
 Set `preserveMultipleSlashes: true` to suppress this behavior.
 
+### strictSlashes
+
+By default on Windows (`platform: 'win32'`), `\` characters in
+the file path to test are treated as `/`, so that a pattern like
+`a/b/c` will match the Windows path `a\b\c`.
+
+Set `strictSlashes: true` to disable this conversion. In this
+mode, `/` in the pattern will only match `/` in the file path,
+and `\` characters will be preserved (and thus will not match
+`/`).
+
+This option has no effect on POSIX platforms.
+
 ### optimizationLevel
 
 A number indicating the level of optimization that should be done
diff --git a/src/index.ts b/src/index.ts
index b20f34f..6730f55 100644
--- a/src/index.ts
+++ b/src/index.ts
@@ -68,6 +68,13 @@ export interface MinimatchOptions {
   flipNegate?: boolean
   /** do not collapse multiple `/` into a single `/` */
   preserveMultipleSlashes?: boolean
+  /**
+   * When set to `true`, pattern `/` only matches `/` in the file path.
+   * That is, the default behavior of converting `\` to `/` in file paths
+   * on Windows platforms is disabled. This is useful on Windows when you
+   * want to distinguish between `/` and `\` as path separators.
+   */
+  strictSlashes?: boolean
   /**
    * A number indicating the level of optimization that should be done
    * to the pattern prior to parsing and using it for matches.
@@ -405,6 +412,7 @@ export class Minimatch {
   platform: Platform
   windowsNoMagicRoot: boolean
   maxGlobstarRecursion: number
+  strictSlashes: boolean
 
   regexp: false | null | MMRegExp
   constructor(pattern: string, options: MinimatchOptions = {}) {
@@ -435,6 +443,7 @@ export class Minimatch {
       options.windowsNoMagicRoot !== undefined ?
         options.windowsNoMagicRoot
       : !!(this.isWindows && this.nocase)
+    this.strictSlashes = !!options.strictSlashes
 
     this.globSet = []
     this.globParts = []
@@ -1424,7 +1433,7 @@ export class Minimatch {
     const options = this.options
 
     // windows: need to use /, not \
-    if (this.isWindows) {
+    if (this.isWindows && !this.strictSlashes) {
       f = f.split('\\').join('/')
     }
 
