diff --git a/README.md b/README.md
index 1a068fa..15cb798 100644
--- a/README.md
+++ b/README.md
@@ -370,6 +370,10 @@ That is, a pattern like `a///b` will match the file path `a/b`.
 
 Set `preserveMultipleSlashes: true` to suppress this behavior.
 
+### strictSlashes
+
+When `strictSlashes: true` is set, `\` characters will not be converted to `/` when matching a path. This allows matching `\` in the pattern against `\` in the path. By default, `\` in paths are converted to `/` on Windows. This option only has an effect on Windows.
+
 ### optimizationLevel
 
 A number indicating the level of optimization that should be done
diff --git a/src/index.ts b/src/index.ts
index b20f34f..018017e 100644
--- a/src/index.ts
+++ b/src/index.ts
@@ -68,6 +68,8 @@ export interface MinimatchOptions {
   flipNegate?: boolean
   /** do not collapse multiple `/` into a single `/` */
   preserveMultipleSlashes?: boolean
+  /** strict slashes: do not convert \ to / on Windows */
+  strictSlashes?: boolean
   /**
    * A number indicating the level of optimization that should be done
    * to the pattern prior to parsing and using it for matches.
@@ -396,6 +398,7 @@ export class Minimatch {
   comment: boolean
   empty: boolean
   preserveMultipleSlashes: boolean
+  strictSlashes: boolean
   partial: boolean
   globSet: string[]
   globParts: string[][]
@@ -424,6 +427,7 @@ export class Minimatch {
       this.pattern = this.pattern.replace(/\\/g, '/')
     }
     this.preserveMultipleSlashes = !!options.preserveMultipleSlashes
+    this.strictSlashes = !!options.strictSlashes
     this.regexp = null
     this.negate = false
     this.nonegate = !!options.nonegate
@@ -1424,7 +1428,7 @@ export class Minimatch {
     const options = this.options
 
     // windows: need to use /, not \
-    if (this.isWindows) {
+    if (this.isWindows && !this.strictSlashes) {
       f = f.split('\\').join('/')
     }
 
