diff --git a/.github/workflows/ci.yml b/.github/workflows/ci.yml
index 19530c3..232fa37 100644
--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -54,3 +54,9 @@ jobs:
 
       - name: Run Tests
         run: npm test -- -c -t0
+
+      - name: Run Lint
+        run: npm run lint
+
+      - name: Run TypeDoc
+        run: npm run typedoc
diff --git a/.github/workflows/typedoc.yml b/.github/workflows/typedoc.yml
index 08af735..ec3d9e5 100644
--- a/.github/workflows/typedoc.yml
+++ b/.github/workflows/typedoc.yml
@@ -23,10 +23,10 @@ jobs:
     steps:
       - name: Checkout
         uses: actions/checkout@v6
-      - name: Use Nodejs ${{ matrix.node-version }}
+      - name: Use Nodejs 22
         uses: actions/setup-node@v6
         with:
-          node-version: 20.x
+          node-version: 22.x
       - name: Install dependencies
         run: npm install
       - name: Generate typedocs
diff --git a/benchmark.js b/benchmark.js
index c10a6ac..1a9c625 100644
--- a/benchmark.js
+++ b/benchmark.js
@@ -1,15 +1,16 @@
-var m = require('./minimatch.js')
-var pattern = '**/*.js'
-var expand = require('brace-expansion')
-var files = expand('x/y/z/{1..1000}.js')
-var start = process.hrtime()
+import { minimatch } from '../dist/esm/index.js'
+import expand from 'brace-expansion'
 
-for (var i = 0; i < 1000; i++) {
-  for (var f = 0; f < files.length; f++) {
-    var res = m(pattern, files[f])
+const pattern = '**/*.js'
+const files = expand('x/y/z/{1..1000}.js')
+const start = process.hrtime()
+
+for (let i = 0; i < 1000; i++) {
+  for (let f = 0; f < files.length; f++) {
+    const res = minimatch(files[f], pattern)
   }
   if (!(i % 10)) process.stdout.write('.')
 }
 console.log('done')
-var dur = process.hrtime(start)
+const dur = process.hrtime(start)
 console.log('%s ms', dur[0] * 1e3 + dur[1] / 1e6)
diff --git a/package.json b/package.json
index 211875c..7977033 100644
--- a/package.json
+++ b/package.json
@@ -35,7 +35,7 @@
     "test": "tap",
     "snap": "tap",
     "format": "prettier --write .",
-    "benchmark": "node benchmark/index.js",
+    "benchmark": "node benchmark.js",
     "typedoc": "typedoc --tsconfig .tshy/esm.json ./src/*.ts",
     "lint": "oxlint --fix src test",
     "postsnap": "npm run lint",
